README
This is an enhanced version of ser2net with the following features:

* multiple connections to the same serial port
* multiple ports of a serial port
* new "http" mode for WebSocket connections
* new "rraw" mode for actively connecting to a remote port instead of
  listening locally

---------------------------------------------------------------------

This is ser2net, a program for allowing network connections to serial
ports.  See the man page for information about using the program.

Note that ser2net supports RFC 2217 (remote control of serial port
parameters), but you must have a compliant client.  The only one I
know if is kermit (http://www.columbia.edu/kermit).

If you want the opposite of ser2net (you want to connect to a "local"
serial port device that is really remote) then Cyclades has provided
a tool for this at http://www.coker.com.au/cyclades.  It is capable
of connecting to ser2net using RFC2217.

If you check this out from git, you won't have all the configure
programs and files, because those are generated.  Do:
   autoreconf -i
to generate everything first.  Then you can run configure.


ChangeLog
2011-12-25  Longshine <longxianghe@gmail.com>

	* dataxfer.c, http.c, http.h, list.c, list.h: Add support for
	multi-connections.

	* devcfg.c, devcfg.h: Add 'tos' to indicate unwrapping TinyOS
	frames.
	
2011-12-21  Longshine <longxianghe@gmail.com>

	* dataxfer.c, http.c, http.h, sha1.c, sha1.h: Add support for 
	HTTP/WebSocket connections.

2011-12-17  Longshine <longxianghe@gmail.com>

	* dataxfer.c: Add "rraw" state to enable ser2net sends serial
	data to a remote endpoint, using original timeout as
	reconnecting interval when connection fails.

2009-11-12  Corey Minyard <minyard@acm.org>

	* ser2net.spec, ser2net.spec.in: Fix the version in the spec
	file to be updated automatically.

2009-11-12  Philip Gwyn

	* dataxfer.c, devcfg.c, devcfg.h, ser2net.8: Add the ability to
	dump trace files in a binary format with timestamps.

2009-11-03  Corey Minyard <minyard@acm.org>

	* configure.in: Move version to 2.7 for release.
	
2009-10-23  Corey Minyard <minyard@acm.org>

	* configure.in: Move to 2.7-rc3.

	* dataxfer.c, devcfg.c, devcfg.h, ser2net.h: Modified the tb=
	directive to work separately from tw= and tr=, so you can do all
	three at the same time independently.

	* readconfig.c: Improved an error report.
	
2009-10-22  Corey Minyard <minyard@acm.org>

	* ser2net.c: Set up routing errors to stderr before reading the
	config so config errors come out properly to stdout.

	* ser2net.8, ser2net.conf: Fix the documentation, it's tw=, not
	tw: for the tracefile stuff.

	* configure.in: Move to 2.7-rc2.

2009-10-15  Corey Minyard <minyard@acm.org>

	* dataxfer.c, devcfg.c, readconfig.c, ser2net.8, ser2net.conf: Add
	the ability to trace data read written to trace files.

2009-07-19  Geoff Simmons <gsimmons@gsimmons.org>

	* ser2net.conf: Some documentation cleanup, added missing docs
	for the timeout parameter.

	* README - Minor doc cleanup.
	
2009-07-15  Corey Minyard <minyard@acm.org>

	* configure.in: Move to version 2.6.
	
2009-06-30  Corey Minyard <minyard@acm.org>

	* dataxfer.c, devcfg.c, devcfg.h, readconfig.h, utils.h: Add the
	a TRACEFILE option to trace all I/O to a file., and separate
	out the defconfig into a separate structure to avoid passing a
	boatload of things to the config.
	
2009-06-30  Sebastian Andrzej Siewior <bigeasy@linutronix.de>

	* controller.c: Fix use after free in controller

          The controller will use its dynamically allocated data after it
	  got free() in error path. What we see in syslog is:

          Jun 30 10:26:38 consrv3 ser2net[3073]: read error for
		controller port: Connection reset by peer
          Jun 30 10:26:39 consrv3 ser2net[3073]: The tcp write for
		controller had error: Bad file descriptor

          The first error is "legal" because the destitnation decided to
	  close its socket a little to early than expected. The second
	  error is allready bad because it tries to use allready
	  deallocated fd. Later we segfault.

2007-11-16  Corey Minyard <minyard@acm.org>

	* configure.in: Move to version 2.5.

2007-11-16  Harry Flick, Corey Minyard <minyard@acm.org>

	* devcfg.c, devcfg.h, dataxfer.c, ser2net.conf, ser2net.8: Allow a
	"NOBREAK" option to disable automatically messing with the break
	operations on the serial port.  Useful for systems with a broken
	break.

2007-11-16  Robert Edmonds <edmonds@debian.org>,Corey Minyard <minyard@acm.org>
	
	* ser2net.c, ser2net.8: Add a '-P' option to specific that a
	pidfile be created.

2007-05-03  Corey Minyard <minyard@acm.org>

	* dataxfer.c: Change the rfc2217 signature response code
	from 0 to 100.  Best guess on what this should be, as the
	RFC doesn't give the correct value and says "text".

2007-04-26  Corey Minyard <minyard@acm.org>

	* dataxfer.c: Disable the Nagle algorithm (enable TCP_NODELAY) on
	the TCP ports to avoid long delays on short pieces of data.

	* configure.in: Move to version 2.4.

2007-03-19  Corey Minyard <minyard@acm.org>

	* dataxfer.c: Use the proper value for modem data notifies sent
	to the client, it was 7, use 107 instead.

2006-12-01  Corey Minyard <minyard@acm.org>

	* readconfig.c: Handle an ending banner that does not have a '\n'
	in the last line.

2006-01-16  Corey Minyard <minyard@acm.org>

	* ser2net.8: Fix the documentation for control ports.

2006-01-09  Corey Minyard <minyard@acm.org>

	* dataxfer.c: Don't kill the session if TIOCCBRK doesn't work.

2005-12-29  Corey Minyard <minyard@acm.org>

	* dataxfer.c, devcfg.c, devcfg.h, ser2net.conf, ser2net.8: Added a
	'\s' banner token for printing out the serial parameters.

	* ser2net.conf: Cleaned up the ugly banner3.

	* Makefile.am: Add ser2net.init to EXTRA_DIST.

	* configure.in: Move to version 2.3.

2005-12-29  Aurelien Jarno <aurel32@debian.org>	
	
	* devcfg.c: Don't include "termio.h", but do include
	"sys/ioctl.h".

2005-10-20  Corey Minyard <minyard@acm.org>

	* dataxfer.c: Fixed a problem with rfc2217 stop bit size setting.
	It was interpreting the value of "3" as 2 stop bits, but 3 is for
	1.5 stop bits, and 2 is for 2 stop bits.

	* controller.c: Fixed some warnings dealing with char/unsigned
	char problems.
	* dataxfer.c: ditto
	* telnet.c: ditto

2005-08-25  Corey Minyard <minyard@acm.org>

	* dataxfer.c: Fix a problem wih EAGAIN just returning and not
	retrying the operation on writing data that has just been read.
	Thanks to Shinlun Hsieh <yngwie_x@yahoo.com> for pointing this out.

2005-07-07  G Goodwill <ggg2sforge at sourceforge.net>

	* dataxfer.c, devcfg.c: Added support for cygwin.

2005-07-07  Rajiv Aaron Manglani <rajav at gentoo.org>

	* configure.in: Make the use of tcp wrappers optional.
	
2005-06-13  Yura Kalinichenko <yuk@iceb.vinnitsa.com>


http.h
/*
 *  ser2nets-http - A program for allowing HTTP connection to serial ports
 *  Copyright (C) 2011  Longshine <longxianghe@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef __HTTP_H__
#define __HTTP_H__

#define HTTP_BUFSIZE    1024

#define HTTP_UNCONNECTED 0
#define HTTP_CONNECTING 1
#define HTTP_CONNECTED 2
#define HTTP_CLOSING 3

#ifdef WORDS_BIGENDIAN
typedef struct ws_hdr_s {
  unsigned char FIN:1;
  unsigned char RSV1:1;
  unsigned char RSV2:1;
  unsigned char RSV3:1;
  unsigned char opcode:4;
  unsigned char MASK:1;
  unsigned char length:7;
} ws_hdr_t;

typedef union ws_length_s {
  unsigned int len64;
  unsigned int len16:16;
  struct {
    unsigned char MASK:1;
    unsigned char len7:7;
  };
} ws_length_t;
#else
typedef struct ws_hdr_s {
  unsigned char opcode:4;
  unsigned char RSV3:1;
  unsigned char RSV2:1;
  unsigned char RSV1:1;
  unsigned char FIN:1;
  unsigned char length:7;
  unsigned char MASK:1;
} ws_hdr_t;

typedef union ws_length_s {
  unsigned char len7:7;
  unsigned int len16:16;
  unsigned int len64;
} ws_length_t;
#endif

typedef struct ws_pdu_s {
  ws_hdr_t *hdr;
  ws_length_t *len;
  unsigned char *mask;
  unsigned char *payload;
  unsigned int index;
  unsigned int left;
  unsigned char buf[HTTP_BUFSIZE];
} ws_frame_t;

typedef struct http_data_s
{
  int state;
  int request_buf_count;
  int request_buf_start;
  unsigned char request_buf[HTTP_BUFSIZE];
  int response_buf_count;
  int response_buf_start;
  unsigned char response_buf[HTTP_BUFSIZE];
  void *cb_data;
  void (*output_ready)(void *cb_data);
  int (*handle_request)(void *cb_data, unsigned char *buf, int size);
  int (*handle_response)(void *cb_data, unsigned char *buf, int size);
  char websocket_key[32];
  ws_frame_t request_frame;
  ws_frame_t response_frame;
} http_data_t;

void http_init(http_data_t *hd, void *cb_data,
               void (*output_ready)(void *),
               int (*handle_request)(void *, unsigned char *, int),
               int (*handle_response)(void *, unsigned char *, int));
char* http_process_request(http_data_t *hd, int fd);
char* http_process_response(http_data_t *hd, unsigned char *buf, int len);

#endif /* __HTTP_H__ */


utils.c
/*
 *  ser2net - A program for allowing telnet connection to serial ports
 *  Copyright (C) 2001  Corey Minyard <minyard@acm.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* This file holds basic utilities used by the ser2net program. */

#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>

#include "utils.h"
#include "selector.h"

extern selector_t *ser2net_sel;

/* Scan for a positive integer, and return it.  Return -1 if the
   integer was invalid. */
int
scan_int(char *str)
{
    int rv = 0;

    if (*str == '\0') {
	return -1;
    }

    for (;;) {
	switch (*str) {
	case '0': case '1': case '2': case '3': case '4':
	case '5': case '6': case '7': case '8': case '9':
	    rv = (rv * 10) + ((*str) - '0');
	    break;

	case '\0':
	    return rv;

	default:
	    return -1;
	}

	str++;
    }

    return rv;
}

/* Scan for a TCP port in the form "[hostname,]x", where the optional
   first part is a resolvable hostname, an IPv4 octet, or an IPv6 address.
   In the absence of a host specification, a wildcard address is used.
   The mandatory second part is the port number or a service name. */
int
scan_tcp_port(char *str, struct addrinfo **rai)
{
    char *strtok_data, *strtok_buffer;
    char *ip;
    char *port;
    struct addrinfo hints, *ai;

    strtok_buffer = strdup(str);
    if (!strtok_buffer)
	return ENOMEM;

    ip = strtok_r(strtok_buffer, ",", &strtok_data);
    port = strtok_r(NULL, "", &strtok_data);
    if (port == NULL) {
	port = ip;
	ip = NULL;
    }

    memset(&hints, 0, sizeof(hints));
    hints.ai_flags = AI_PASSIVE;
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    if (getaddrinfo(ip, port, &hints, &ai)) {
	free(strtok_buffer);
	return EINVAL;
    }

    free(strtok_buffer);
    if (*rai)
	freeaddrinfo(*rai);
    *rai = ai;
    return 0;
}

void
check_ipv6_only(int family, struct sockaddr *addr, int fd)
{
    if ((family == AF_INET6)
	&& IN6_IS_ADDR_UNSPECIFIED(&(((struct sockaddr_in6 *) addr)->sin6_addr)))
    {
	int null = 0;

	setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &null, sizeof(null));
    }
}

int *
open_socket(struct addrinfo *ai, void (*readhndlr)(int, void *), void *data,
	    unsigned int *nr_fds)
{
    struct addrinfo *rp;
    int optval = 1;
    int family = AF_INET6; /* Try IPV6 first, then IPV4. */
    int *fds;
    unsigned int curr_fd = 0;
    unsigned int max_fds = 0;

    for (rp = ai; rp != NULL; rp = rp->ai_next)
	max_fds++;

    fds = malloc(sizeof(int) * max_fds);
    if (!fds)
	return NULL;

  restart:
    for (rp = ai; rp != NULL; rp = rp->ai_next) {
	if (family != rp->ai_family)
	    continue;

	fds[curr_fd] = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
	if (fds[curr_fd] == -1)
	    continue;

	if (fcntl(fds[curr_fd], F_SETFL, O_NONBLOCK) == -1)
	    goto next;

	if (setsockopt(fds[curr_fd], SOL_SOCKET, SO_REUSEADDR,
		       (void *)&optval, sizeof(optval)) == -1)
	    goto next;

	check_ipv6_only(rp->ai_family, rp->ai_addr, fds[curr_fd]);

	if (bind(fds[curr_fd], rp->ai_addr, rp->ai_addrlen) != 0)
	    goto next;

	if (listen(fds[curr_fd], 1) != 0)
	    goto next;

	sel_set_fd_handlers(ser2net_sel, fds[curr_fd], data,
			    readhndlr, NULL, NULL);
	sel_set_fd_read_handler(ser2net_sel, fds[curr_fd],
				SEL_FD_HANDLER_ENABLED);
	curr_fd++;
	continue;

      next:
	close(fds[curr_fd]);
    }
    if (family == AF_INET6) {
	family = AF_INET;
	goto restart;
    }

    if (curr_fd == 0) {
	free(fds);
	fds = NULL;
    }
    *nr_fds = curr_fd;
    return fds;
}

int *
connect_socket(struct addrinfo *ai, unsigned int *nr_fds)
{
    struct addrinfo *rp;
    int family = AF_INET6; /* Try IPV6 first, then IPV4. */
    int *fds;
    unsigned int curr_fd = 0;
    unsigned int max_fds = 0;

    for (rp = ai; rp != NULL; rp = rp->ai_next)
	max_fds++;

    fds = malloc(sizeof(int) * max_fds);
    if (!fds)
	return NULL;

  restart:
    for (rp = ai; rp != NULL; rp = rp->ai_next) {
	if (family != rp->ai_family)
	    continue;

	fds[curr_fd] = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
	if (fds[curr_fd] == -1)
	    continue;

	check_ipv6_only(rp->ai_family, rp->ai_addr, fds[curr_fd]);

	if (connect(fds[curr_fd], rp->ai_addr, rp->ai_addrlen) != 0)
	    goto next;

	curr_fd++;
	continue;

      next:
	close(fds[curr_fd]);
    }
    if (family == AF_INET6) {
	family = AF_INET;
	goto restart;
    }

    if (curr_fd == 0) {
	free(fds);
	fds = NULL;
    }
    *nr_fds = curr_fd;
    return fds;
}

int
write_full(int fd, char *data, size_t count)
{
    size_t written;

 restart:
    while ((written = write(fd, data, count)) > 0) {
	data += written;
	count -= written;
    }
    if (written < 0) {
	if (errno == EAGAIN)
	    goto restart;
	return -1;
    }
    return 0;
}

void
write_ignore_fail(int fd, char *data, size_t count)
{
    ssize_t written;

    while ((written = write(fd, data, count)) > 0) {
	data += written;
	count -= written;
    }
}


devcfg.c
/*
 *  ser2net - A program for allowing telnet connection to serial ports
 *  Copyright (C) 2001  Corey Minyard <minyard@acm.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* This code handles generating the configuration for the serial port. */

#include <unistd.h>
#include <termios.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "devcfg.h"
#include "utils.h"

#ifdef __CYGWIN__
void cfmakeraw(struct termios *termios_p) {
    termios_p->c_iflag &= ~(IGNBRK|BRKINT|PARMRK|ISTRIP|INLCR|IGNCR|ICRNL|IXON);
    termios_p->c_oflag &= ~OPOST;
    termios_p->c_lflag &= ~(ECHO|ECHONL|ICANON|ISIG|IEXTEN);
    termios_p->c_cflag &= ~(CSIZE|PARENB);
    termios_p->c_cflag |= CS8;
}
#endif

/* Initialize a serial port control structure for the first time.
   This should only be called when the port is created.  It sets the
   port to the default 9600N81. */
void
devinit(struct termios *termctl)
{
    cfmakeraw(termctl);
    cfsetospeed(termctl, B9600);
    cfsetispeed(termctl, B9600);
    termctl->c_cflag &= ~(CSTOPB);
    termctl->c_cflag &= ~(CSIZE);
    termctl->c_cflag |= CS8;
    termctl->c_cflag &= ~(PARENB);
    termctl->c_cflag &= ~(CLOCAL);
    termctl->c_cflag &= ~(HUPCL);
    termctl->c_cflag |= CREAD;
    termctl->c_cflag &= ~(CRTSCTS);
    termctl->c_iflag &= ~(IXON | IXOFF | IXANY);
    termctl->c_iflag |= IGNBRK;
}

void
devinitinfo(dev_info_t *dinfo)
{
    devinit(&(dinfo->termctl));

    dinfo->allow_2217 = 0;
    dinfo->disablebreak = 0;
    dinfo->banner = NULL;
    dinfo->is_tos_frame = 0;
}

/* Configure a serial port control structure based upon input strings
   in instr.  These strings are described in the man page for this
   program. */
int
devconfig(char *instr, dev_info_t *dinfo)
{
    char *str;
    char *pos;
    char *strtok_data;
    int  rv = 0;
    struct termios *termctl = &dinfo->termctl;

    str = strdup(instr);
    if (str == NULL) {
	return -1;
    }

    dinfo->allow_2217 = 0;
    dinfo->disablebreak = 0;
    dinfo->banner = NULL;
    dinfo->trace_read.file = NULL;
    dinfo->trace_read.hexdump = 0;
    dinfo->trace_read.timestamp = 0;
    dinfo->trace_write.file = NULL;
    dinfo->trace_write.hexdump = 0;
    dinfo->trace_write.timestamp = 0;
    dinfo->trace_both.file = NULL;
    dinfo->trace_both.hexdump = 0;
    dinfo->trace_both.timestamp = 0;
    pos = strtok_r(str, ", \t", &strtok_data);
    while (pos != NULL) {
	if (strcmp(pos, "300") == 0) {
	    cfsetospeed(termctl, B300);
	    cfsetispeed(termctl, B300);
	} else if (strcmp(pos, "1200") == 0) {
	    cfsetospeed(termctl, B1200);
	    cfsetispeed(termctl, B1200);
	} else if (strcmp(pos, "2400") == 0) {
	    cfsetospeed(termctl, B2400);
	    cfsetispeed(termctl, B2400);
	} else if (strcmp(pos, "4800") == 0) {
	    cfsetospeed(termctl, B4800);
	    cfsetispeed(termctl, B4800);
	} else if (strcmp(pos, "9600") == 0) {
	    cfsetospeed(termctl, B9600);
	    cfsetispeed(termctl, B9600);
	} else if (strcmp(pos, "19200") == 0) {
	    cfsetospeed(termctl, B19200);
	    cfsetispeed(termctl, B19200);
	} else if (strcmp(pos, "38400") == 0) {
	    cfsetospeed(termctl, B38400);
	    cfsetispeed(termctl, B38400);
	} else if (strcmp(pos, "57600") == 0) {
	    cfsetospeed(termctl, B57600);
	    cfsetispeed(termctl, B57600);
	} else if (strcmp(pos, "115200") == 0) {
	    cfsetospeed(termctl, B115200);
	    cfsetispeed(termctl, B115200);
	} else if (strcmp(pos, "1STOPBIT") == 0) {
	    termctl->c_cflag &= ~(CSTOPB);
	} else if (strcmp(pos, "2STOPBITS") == 0) {
	    termctl->c_cflag |= CSTOPB;
	} else if (strcmp(pos, "7DATABITS") == 0) {
	    termctl->c_cflag &= ~(CSIZE);
	    termctl->c_cflag |= CS7;
	} else if (strcmp(pos, "8DATABITS") == 0) {
	    termctl->c_cflag &= ~(CSIZE);
	    termctl->c_cflag |= CS8;
	} else if (strcmp(pos, "NONE") == 0) {
	    termctl->c_cflag &= ~(PARENB);
	} else if (strcmp(pos, "EVEN") == 0) {
	    termctl->c_cflag |= PARENB;
	    termctl->c_cflag &= ~(PARODD);
	} else if (strcmp(pos, "ODD") == 0) {
	    termctl->c_cflag |= PARENB | PARODD;
        } else if (strcmp(pos, "XONXOFF") == 0) {
            termctl->c_iflag |= (IXON | IXOFF | IXANY);
            termctl->c_cc[VSTART] = 17;
            termctl->c_cc[VSTOP] = 19;      
        } else if (strcmp(pos, "-XONXOFF") == 0) {
            termctl->c_iflag &= ~(IXON | IXOFF | IXANY);
        } else if (strcmp(pos, "RTSCTS") == 0) {
            termctl->c_cflag |= CRTSCTS;  
        } else if (strcmp(pos, "-RTSCTS") == 0) {
            termctl->c_cflag &= ~CRTSCTS;
        } else if (strcmp(pos, "LOCAL") == 0) {
            termctl->c_cflag |= CLOCAL;  
        } else if (strcmp(pos, "-LOCAL") == 0) {
            termctl->c_cflag &= ~CLOCAL;
        } else if (strcmp(pos, "HANGUP_WHEN_DONE") == 0) {
            termctl->c_cflag |= HUPCL;  
        } else if (strcmp(pos, "-HANGUP_WHEN_DONE") == 0) {
            termctl->c_cflag &= ~HUPCL;
        } else if (strcmp(pos, "remctl") == 0) {
	    dinfo->allow_2217 = 1;
	} else if (strcmp(pos, "NOBREAK") == 0) {
	    dinfo->disablebreak = 1;
	} else if (strcmp(pos, "hexdump") == 0 ||
	           strcmp(pos, "-hexdump") == 0) {
	    dinfo->trace_read.hexdump = (*pos != '-');
	    dinfo->trace_write.hexdump = (*pos != '-');
	    dinfo->trace_both.hexdump = (*pos != '-');
	} else if (strcmp(pos, "timestamp") == 0 ||
	           strcmp(pos, "-timestamp") == 0) {
	    dinfo->trace_read.timestamp = (*pos != '-');
	    dinfo->trace_write.timestamp = (*pos != '-');
	    dinfo->trace_both.timestamp = (*pos != '-');
	} else if (strcmp(pos, "tr-hexdump") == 0 ||
	           strcmp(pos, "-tr-hexdump") == 0) {
	    dinfo->trace_read.hexdump = (*pos != '-');
	} else if (strcmp(pos, "tr-timestamp") == 0 ||
	           strcmp(pos, "-tr-timestamp") == 0) {
	    dinfo->trace_read.timestamp = (*pos != '-');
	} else if (strcmp(pos, "tw-hexdump") == 0 ||
	           strcmp(pos, "-tw-hexdump") == 0) {
	    dinfo->trace_write.hexdump = (*pos != '-');
	} else if (strcmp(pos, "tw-timestamp") == 0 ||
	           strcmp(pos, "-tw-timestamp") == 0) {
	    dinfo->trace_write.timestamp = (*pos != '-');
	} else if (strcmp(pos, "tb-hexdump") == 0 ||
	           strcmp(pos, "-tb-hexdump") == 0) {
	    dinfo->trace_both.hexdump = (*pos != '-');
	} else if (strcmp(pos, "tb-timestamp") == 0 ||
	           strcmp(pos, "-tb-timestamp") == 0) {
	    dinfo->trace_both.timestamp = (*pos != '-');
	} else if (strncmp(pos, "tr=", 3) == 0) {
	    /* trace read, data from the port to the socket */
	    dinfo->trace_read.file = find_tracefile(pos + 3);
	} else if (strncmp(pos, "tw=", 3) == 0) {
	    /* trace write, data from the socket to the port */
	    dinfo->trace_write.file = find_tracefile(pos + 3);
	} else if (strncmp(pos, "tb=", 3) == 0) {
	    /* trace both directions. */
	    dinfo->trace_both.file = find_tracefile(pos + 3);
	} else if ((dinfo->banner = find_banner(pos))) {
	    /* It's a banner to display at startup, it's already set. */
	} else if (strcmp(pos, "tos") == 0) {
	    dinfo->is_tos_frame = 1;
	} else {
	    /* fprintf( stderr, "Unknown token %s\n", pos );     */
	    rv = -1;
	    goto out;
	}

	pos = strtok_r(NULL, ", \t", &strtok_data);
    }

out:
    free(str);
    return rv;
}

static char *
baud_string(int speed)
{
    char *str;
    switch (speed) {
    case B300: str = "300"; break;
    case B1200: str = "1200"; break;
    case B2400: str = "2400"; break;
    case B4800: str = "4800"; break;
    case B9600: str = "9600"; break;
    case B19200: str = "19200"; break;
    case B38400: str = "38400"; break;
    case B57600: str = "57600"; break;
    case B115200: str = "115200"; break;
    default: str = "unknown speed";
    }
    return str;
}

void
serparm_to_str(char *str, int strlen, struct termios *termctl)
{
    speed_t speed = cfgetospeed(termctl);
    int     stopbits = termctl->c_cflag & CSTOPB;
    int     databits = termctl->c_cflag & CSIZE;
    int     parity_enabled = termctl->c_cflag & PARENB;
    int     parity = termctl->c_cflag & PARODD;
    char    *sstr;
    char    pchar, schar, dchar;

    sstr = baud_string(speed);

    if (stopbits) 
	schar = '2';
    else
	schar = '1';

    switch (databits) {
    case CS7: dchar = '7'; break;
    case CS8: dchar = '8'; break;
    default: dchar = '?';
    }

    if (parity_enabled) {
	if (parity) {
	    pchar = 'O';
	} else {
	    pchar = 'E';
	}
    } else {
	pchar = 'N';
    }

    snprintf(str, strlen, "%s %c%c%c", sstr, pchar, dchar, schar);
}

/* Send the serial port device configuration to the control port. */
void
show_devcfg(struct controller_info *cntlr, struct termios *termctl)
{
    speed_t speed = cfgetospeed(termctl);
    int     stopbits = termctl->c_cflag & CSTOPB;
    int     databits = termctl->c_cflag & CSIZE;
    int     parity_enabled = termctl->c_cflag & PARENB;
    int     parity = termctl->c_cflag & PARODD;
    int     xon = termctl->c_iflag & IXON;
    int     xoff = termctl->c_iflag & IXOFF;
    int     xany = termctl->c_iflag & IXANY;
    int     flow_rtscts = termctl->c_cflag & CRTSCTS;
    int     clocal = termctl->c_cflag & CLOCAL;
    int     hangup_when_done = termctl->c_cflag & HUPCL;
    char    *str;

    str = baud_string(speed);
    controller_output(cntlr, str, strlen(str));
    controller_output(cntlr, " ", 1);

    if (xon && xoff && xany) {
      controller_output(cntlr, "XONXOFF ", 8);
    }      
    
    if (flow_rtscts) {
      controller_output(cntlr, "RTSCTS ", 7);
    }

    if (clocal) {
      controller_output(cntlr, "LOCAL ", 6);
    }

    if (hangup_when_done) {
      controller_output(cntlr, "HANGUP_WHEN_DONE ", 17);
    }

    if (stopbits) {
	str = "2STOPBITS";
    } else {
	str = "1STOPBIT";
    }
    controller_output(cntlr, str, strlen(str));
    controller_output(cntlr, " ", 1);

    switch (databits) {
    case CS7: str = "7DATABITS"; break;
    case CS8: str = "8DATABITS"; break;
    default: str = "unknown databits";
    }
    controller_output(cntlr, str, strlen(str));
    controller_output(cntlr, " ", 1);

    if (parity_enabled) {
	if (parity) {
	    str = "ODD";
	} else {
	    str = "EVEN";
	}
    } else {
	str = "NONE";
    }
    controller_output(cntlr, str, strlen(str));
}

int
setdevcontrol(char *instr, int fd)
{
    int rv = 0;
    char *str;
    char *pos;
    int status;
    char *strtok_data;

    str = malloc(strlen(instr) + 1);
    if (str == NULL) {
	return -1;
    }

    strcpy(str, instr);

    pos = strtok_r(str, " \t", &strtok_data);
    while (pos != NULL) {
       if (strcmp(pos, "RTSHI") == 0) {
           ioctl(fd, TIOCMGET, &status);
           status |= TIOCM_RTS;
           ioctl(fd, TIOCMSET, &status);
       } else if (strcmp(pos, "RTSLO") == 0) {
           ioctl(fd, TIOCMGET, &status);
           status &= ~TIOCM_RTS;
           ioctl(fd, TIOCMSET, &status);
       } else if (strcmp(pos, "DTRHI") == 0) {
           ioctl(fd, TIOCMGET, &status);
           status |= TIOCM_DTR;
           ioctl(fd, TIOCMSET, &status);
       } else if (strcmp(pos, "DTRLO") == 0) {
           ioctl(fd, TIOCMGET, &status);
           status &= ~TIOCM_DTR;               /* AKA drop DTR */
           ioctl(fd, TIOCMSET, &status);
	} else {
	    rv = -1;
	    goto out;
	}

	pos = strtok_r(NULL, " \t", &strtok_data);
    }

out:
    free(str);
    return rv;
}

void
show_devcontrol(struct controller_info *cntlr, int fd)
{
    char *str;
    int  status;

    ioctl(fd, TIOCMGET, &status);

    if (status & TIOCM_RTS) {
	str = "RTSHI";
    } else {
	str = "RTSLO";
    }
    controller_output(cntlr, str, strlen(str));
    controller_output(cntlr, " ", 1);

    if (status & TIOCM_DTR) {
	str = "DTRHI";
    } else {
	str = "DTRLO";
    }
    controller_output(cntlr, str, strlen(str));
    controller_output(cntlr, " ", 1);
}
